src/Core/Form/IdentifiableObject/DataHandler/CategoryFormDataHandler.php

```diff
@@ -58,39 +64,54 @@ final class CategoryFormDataHandler implements FormDataHandlerInterface
      */
     private $categoryMenuThumbnailUploader;
 
+    /**
+     * @var MenuThumbnailAvailableKeyProvider
+     */
+    private $menuThumbnailAvailableKeyProvider;
+
     /**
      * @param CommandBusInterface $commandBus
      * @param ImageUploaderInterface $categoryCoverUploader
      * @param ImageUploaderInterface $categoryThumbnailUploader
      * @param ImageUploaderInterface $categoryMenuThumbnailUploader
+     * @param MenuThumbnailAvailableKeyProvider $menuThumbnailAvailableKeyProvider
      */
     public function __construct(
         CommandBusInterface $commandBus,
         ImageUploaderInterface $categoryCoverUploader,
         ImageUploaderInterface $categoryThumbnailUploader,
-        ImageUploaderInterface $categoryMenuThumbnailUploader
+        ImageUploaderInterface $categoryMenuThumbnailUploader,
+        MenuThumbnailAvailableKeyProvider $menuThumbnailAvailableKeyProvider
     ) {
         $this->commandBus = $commandBus;
         $this->categoryCoverUploader = $categoryCoverUploader;
         $this->categoryThumbnailUploader = $categoryThumbnailUploader;
         $this->categoryMenuThumbnailUploader = $categoryMenuThumbnailUploader;
+        $this->menuThumbnailAvailableKeyProvider = $menuThumbnailAvailableKeyProvider;
     }
 
     /**
      * {@inheritdoc}
      */
     public function create(array $data)
     {
+        if (!isset($data['menu_thumbnail_images']) && count($data['menu_thumbnail_images']) > count(MenuThumbnailId::ALLOWED_ID_VALUES)) {
+            throw new MenuThumbnailsLimitException('Maximum number of menu thumbnails exceeded for new category');
+        }
```

Comment:
Seems like this check is very related to the Category business domain (or at least technical domain), so this check should be done in the handler. It's too much responsibility to ask the form data handler to do this.

The fact that `MenuThumbnailAvailableKeyProvider` is injected here is already a sign that too many things are done here (it's not the only one but that's another question 😅)